client/src/services/link.js

```diff
@@ -45,85 +45,96 @@ export function LinkProvider(props) {
-      await postLinkEvent({
-        userId,
-        link_session_id,
-        type: 'success',
-      });
+  const generateLinkConfigs = useCallback(
+    async (isOauth, userId, itemId, oauthToken) => {
+      // The config creation will be bypassed if configs already exist for
+      // that User or Item.
+
+      if (
+        !(itemId != null && !hasRequested.current.byItem[itemId]) &&
+        !(userId != null && !hasRequested.current.byUser[userId])
+      ) {
+        return;
+      }
+      const isUpdate = itemId != null;
+
       if (isUpdate) {
-        await setItemState(itemId, 'good');
-        getItemById(itemId, true);
+        hasRequested.current.byItem[itemId] = true;
+      } else {
+        hasRequested.current.byUser[userId] = true;
+      }
+      let token;
+      if (!isOauth) {
+        const linkTokenResponse = await getLinkToken({ itemId, userId });
+        token = await linkTokenResponse.data.link_token;
       } else {
-        await exchangeToken(publicToken, institution, accounts, userId);
-        getItemsByUser(userId, true);
+        token = oauthToken;
       }
 
-      if (window.location.pathname === '/') {
+      const onSuccess = async (
+        publicToken,
+        { institution, accounts, link_session_id }
+      ) => {
+        logEvent('onSuccess', { institution, accounts, link_session_id });
+        await postLinkEvent({
+          userId,
+          link_session_id,
+          type: 'success',
+        });
+        if (isUpdate) {
+          await setItemState(itemId, 'good');
+          getItemById(itemId, true);
+        } else {
+          await exchangeToken(publicToken, institution, accounts, userId);
+          getItemsByUser(userId, true);
+        }
+
         window.location.href = `/user/${userId}/items`;
-      }
-    };
+      };
 
-    const onExit = async (
-      error,
-      { institution, link_session_id, request_id }
-    ) => {
-      logEvent('onExit', {
+      const onExit = async (
         error,
-        institution,
-        link_session_id,
-        request_id,
-      });
-      const eventError = error || {};
-      await postLinkEvent({
-        userId,
-        link_session_id,
-        request_id,
-        type: 'exit',
-        ...eventError,
-      });
-      if (error != null && error.error_code === 'INVALID_LINK_TOKEN') {
-        await generateLinkConfigs(userId, itemId);
+        { institution, link_session_id, request_id }
+      ) => {
+        logEvent('onExit', {
+          error,
+          institution,
+          link_session_id,
+          request_id,
+        });
+        const eventError = error || {};
+        await postLinkEvent({
+          userId,
+          link_session_id,
+          request_id,
+          type: 'exit',
+          ...eventError,
+        });
+        if (error != null && error.error_code === 'INVALID_LINK_TOKEN') {
+          await generateLinkConfigs(userId, itemId);
+        }
+      };
```

Comment:
does `onError` and `onExit` handle OAuth? I believe there were cases where the oauth redirect just shows a blank white screen if there is an error of if you exit link manually

Reply:
The onExit handles OAuth; the blank page we got was because of the error that occurred within the onSuccess function, which is now fixed.  Is this acceptable or do I need to build in error handling within the onSuccess function?

Reply:
In the onSuccess it would be good to catch the error if we fail to send the public_token to the server. But other than that, no, i think this is good